File: python/knetcat.py
import socket
import sys, getopt
import select, signal
#import threading
#from time import sleep

server = 1
port = 6969
host = "0.0.0.0"
#stoprequest = threading.Event()
rt = ""


def usage():
	print "Server mode usage: knetcat -l <ip address> <port>"
	print "Client mode usage: knetcat <ip address> <port>"

def signal_handler(signal, frame):
	#stoprequest.set()
	#rt.join(stoprequest)
	sys.exit(0)

def serverrun(host, port):
	s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
	s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
	s.bind((host, port))
	s.listen(5)
	if select.select([sys.stdin,],[],[],0.0)[0]:
		c, addr = s.accept()
		for line in sys.stdin.readlines():
			c.send(line)
	else:
		c, addr = s.accept()
		while True:
			data = c.recv(64)
			if not data: break
			sys.stdout.write(data)
	try:
		c.shutdown(1)
	except socket.error as ser:
		print "Socket error, continuing"
	c.close()

def recv_thread(s):
	while not stoprequest.isSet():
		data = s.recv(64)
		sys.stdout.write(data)
		sleep(0.4)

def clientrun(host, port):
	if select.select([sys.stdin,],[],[],0.0)[0]:
		s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
		s = socket.socket()
		s.connect((host, port))
		for line in sys.stdin.readlines():
			s.send(line)
		s.shutdown(1)
		s.close()
	elif not sys.stdout.isatty():
		s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
		s = socket.socket()
		s.connect((host, port))
		while True:
			data = s.recv(64)
			if not data: break
			sys.stdout.write(data)
		s.shutdown(1)
		s.close()
	elif sys.stdout.isatty():
		buf = ""
		s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
		s = socket.socket()
		s.connect((host, port))
		#rt = threading.Thread(target=recv_thread, args=(s,)).start()
		while True:
			buf = sys.stdin.readline()
			if buf != "":
				s.send(buf)
			while True:
				data = s.recv(1024)
				if not data: break
				
				sys.stdout.write(data)
		s.close()

def main():
	if server == 1:
		serverrun(host, port)
	elif server == 0:
		clientrun(host, port)

argv = sys.argv[1:]
try:
	opts, args = getopt.getopt(argv,"l:h:")
except getopt.GetoptError as geter:
	print geter

if not '-l' in sys.argv[1:]:
	server = 0
try: 
	for opt, arg in opts:
		if opt == '-l':
			server = 1
except NameError as ner:
	print ner
	
if '-h' in sys.argv[1:]:
	usage()
	exit(0)
	
if sys.argv[1:]:
	port = int(sys.argv.pop())
	host = sys.argv.pop()

signal.signal(signal.SIGINT, signal_handler)

main()

File: c/knetcat.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>
#include <netinet/in.h>
#define PORT 6969

int main (int argc, char *argv[])
{
	char inbuf[16384];
	//while (fgets(inbuf, 128, stdin) != NULL ) {;
		//send(csock, inbuf, 16384, 0);
	//	printf("%s", inbuf);
	//}
	FILE *cmand;
        int sock, csock, recv_thread;
        int status = 1;
        struct sockaddr_in address;
        int addrlen = sizeof(address);
        int opt = 1;
        char rcmand[128];
        char sysout[16384];
        if ((sock = socket(AF_INET, SOCK_STREAM, 0)) == 0) {
                perror("Could not connect server.");
                exit(EXIT_FAILURE);
        }
        if (setsockopt(sock, SOL_SOCKET, SO_REUSEADDR | SO_REUSEPORT, &opt, sizeof(opt))) {
                perror("setsockopt");
                exit(EXIT_FAILURE);
        }
        address.sin_family = AF_INET;
        address.sin_addr.s_addr = INADDR_ANY;
        address.sin_port = htons( PORT );

        if (bind(sock, (struct sockaddr *)&address, sizeof(address))<0) {
                perror("Could not bind server");
                exit(EXIT_FAILURE);
        }
        listen(sock, 1);
        if ((csock = accept(sock, (struct sockaddr *)&address,(socklen_t*)&addrlen))<0)
	{
                perror("Could not accept client");
                exit(EXIT_FAILURE);
	}
	printf("shit");
	//	char inbuf[16384];
	while (fgets(inbuf, 128, stdin) != NULL ) {;
		printf("check\n");
		send(csock, inbuf, 16384, 0);
		printf("%s", inbuf);
	}
	close(csock);
}
